fix deepspeed special test `test_deepspeed_stage_3_save_warning`

## 🐛 Bug

<!-- A clear and concise description of what the bug is. -->

### To Reproduce

Run on current master branch the following command:

`SPECIAL_PATTERN=test_deepspeed_stage_3_save_warning bash tests/special_tests.sh
`

Error:

```bash

______________________________ test_deepspeed_stage_3_save_warning _______________________________

tmpdir = local('/tmp/pytest-of-adrian/pytest-1376/test_deepspeed_stage_3_save_wa0')

    @RunIf(min_gpus=2, deepspeed=True, special=True)
    def test_deepspeed_stage_3_save_warning(tmpdir):
        """
        Test to ensure that DeepSpeed Stage 3 gives a warning when saving.
        """
        model = BoringModel()
        trainer = Trainer(
            default_root_dir=tmpdir, plugins=[DeepSpeedPlugin(stage=3)], gpus=2, fast_dev_run=True, precision=16
        )
        trainer.fit(model)
        checkpoint_path = os.path.join(tmpdir, "model.pt")
        with pytest.warns(UserWarning, match="each worker will save a shard of the checkpoint within a directory."):
>           trainer.save_checkpoint(checkpoint_path)
E           Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) was emitted. The list of emitted warnings is: [].

tests/plugins/test_deepspeed_plugin.py:432: Failed


```

### Expected behavior

<!-- FILL IN -->

### Environment

* CUDA:
        - GPU:
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.9.0+cu111
        - pytorch-lightning: 1.5.0dev
        - tqdm:              4.60.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.8
        - version:           #64-Ubuntu SMP Wed Dec 9 08:16:25 UTC 2020


### Additional context

In our GPU CI as well, check section for "special tests"

--- a/pytorch_lightning/plugins/training_type/deepspeed.py
+++ b/pytorch_lightning/plugins/training_type/deepspeed.py
@@ -669,7 +669,6 @@ class DeepSpeedPlugin(DDPPlugin):
             filepath: write-target file's path
         """
         if self.zero_stage_3 and self._multi_device and self.is_global_zero:
-            # todo (sean): Add link to docs once docs are merged.
             warning_cache.warn(
                 "When saving the DeepSpeed Stage 3 checkpoint, "
                 "each worker will save a shard of the checkpoint within a directory. "

--- a/tests/plugins/test_deepspeed_plugin.py
+++ b/tests/plugins/test_deepspeed_plugin.py
@@ -409,9 +409,13 @@ def test_deepspeed_stage_3_save_warning(tmpdir):
     )
     trainer.fit(model)
     checkpoint_path = os.path.join(tmpdir, "model.pt")
+    with pytest.warns(UserWarning) as record:
+        # both ranks need to call save checkpoint
+        trainer.save_checkpoint(checkpoint_path)
     if trainer.is_global_zero:
-        with pytest.warns(UserWarning, match="each worker will save a shard of the checkpoint within a directory."):
-            trainer.save_checkpoint(checkpoint_path)
+        assert len(record) == 1
+        match = "each worker will save a shard of the checkpoint within a directory."
+        assert match in str(record[0].message)
 
 
 @RunIf(min_gpus=1, deepspeed=True, special=True)

--- a/tests/plugins/test_deepspeed_plugin.py
+++ b/tests/plugins/test_deepspeed_plugin.py
@@ -420,7 +420,7 @@ def test_deepspeed_fp32_works(tmpdir):
 @RunIf(min_gpus=2, deepspeed=True, special=True)
 def test_deepspeed_stage_3_save_warning(tmpdir):
     """
-    Test to ensure that DeepSpeed Stage 3 gives a warning when saving.
+    Test to ensure that DeepSpeed Stage 3 gives a warning when saving on rank zero.
     """
     model = BoringModel()
     trainer = Trainer(
@@ -428,8 +428,9 @@ def test_deepspeed_stage_3_save_warning(tmpdir):
     )
     trainer.fit(model)
     checkpoint_path = os.path.join(tmpdir, "model.pt")
-    with pytest.warns(UserWarning, match="each worker will save a shard of the checkpoint within a directory."):
-        trainer.save_checkpoint(checkpoint_path)
+    if trainer.is_global_zero:
+        with pytest.warns(UserWarning, match="each worker will save a shard of the checkpoint within a directory."):
+            trainer.save_checkpoint(checkpoint_path)
 
 
 @RunIf(min_gpus=1, deepspeed=True, special=True)

--- a/tests/plugins/test_deepspeed_plugin.py
+++ b/tests/plugins/test_deepspeed_plugin.py
@@ -402,9 +402,7 @@ def test_deepspeed_fp32_works(tmpdir):
 
 @RunIf(min_gpus=2, deepspeed=True, special=True)
 def test_deepspeed_stage_3_save_warning(tmpdir):
-    """
-    Test to ensure that DeepSpeed Stage 3 gives a warning when saving on rank zero.
-    """
+    """Test to ensure that DeepSpeed Stage 3 gives a warning when saving on rank zero."""
     model = BoringModel()
     trainer = Trainer(
         default_root_dir=tmpdir, plugins=[DeepSpeedPlugin(stage=3)], gpus=2, fast_dev_run=True, precision=16

--- a/tests/trainer/loops/test_training_loop.py
+++ b/tests/trainer/loops/test_training_loop.py
@@ -191,7 +191,7 @@ def test_prepare_outputs(tmpdir):
 
 
 def test_batch_loop_releases_loss(tmpdir):
-    """Test that loss/graph is released so that it can be garbage collected before the next training step"""
+    """Test that loss/graph is released so that it can be garbage collected before the next training step."""
 
     class TestModel(BoringModel):
         def training_step(self, batch, batch_idx):

